### ISSUE TITLE
Support for overriding only individual socket functions?

### ISSUE THREAD

**Issue by timwoj** [c-ares/c-ares#441]
I'm working on a project where I need to know when the socket for a resolver closes. I see that I can override a number of internal methods on the channel using `ares_set_socket_functions`. Looking through the rest of the code it looks like you're required to override all of them, but I really only need to override `aclose`.

I considered opening a PR that fixes all of the places in `ares_process.c` that check for `channel->sock_funcs` to be null and to also check the corresponding method called in the block under the check, but one spot is confusing me slightly. `configure_socket` bails out entirely if `sock_funcs` is not null, but it doesn't appear to use any of them. Is that an unnecessary check, or is there something deeper happening there that I'm not seeing?

**Comment by bradh352** [comment]
sounds like you just need to register ARES_OPT_SOCK_STATE_CB and wait for the callback to set both readable and writeable to 0, or will that not do something you're looking for.  I'd have to look at what you're talking about regarding sock_funcs in configure_socket ... i'll try to do that in a couple of days.

**Comment by bradh352** [comment]
regarding your comment about configure_socket, that appears to mostly call setsockopt() which if you have your own networking stack, can't be called, nor the bind() call in there, so its right for it to be bypassed.

That said, it looks like after that function is called, another setsockopt() is called, so that's actually bad.

Truthfully, this should have been implemented in such a way that everything uses the socket callbacks, and just by default they point to the default implementation.  It looks like this feature was just sort of hacked in.

That said, on your original goal, like stated previously, you probably should go the route I suggested instead and use the socket state callbacks.
